File: android/app/src/main/java/com/carelog/ui/vitals/SpO2ViewModel.kt
package com.carelog.ui.vitals

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.carelog.auth.AuthRepository
import com.carelog.fhir.client.ObservationType
import com.carelog.fhir.models.FhirObservation
import com.carelog.fhir.repository.LocalFhirRepository
import com.carelog.voice.VoiceAcknowledgementPlayer
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.time.Instant
import java.util.UUID
import javax.inject.Inject

/**
 * ViewModel for SpO2 (oxygen saturation) logging screen.
 */
@HiltViewModel
class SpO2ViewModel @Inject constructor(
    private val authRepository: AuthRepository,
    private val localFhirRepository: LocalFhirRepository,
    private val voicePlayer: VoiceAcknowledgementPlayer
) : ViewModel() {

    private val _uiState = MutableStateFlow(SpO2UiState())
    val uiState: StateFlow<SpO2UiState> = _uiState.asStateFlow()

    fun updateValue(value: String) {
        _uiState.update { state ->
            val intValue = value.toIntOrNull()
            val error = validateValue(intValue)
            val showWarning = intValue != null && intValue in 90..94
            state.copy(
                value = value,
                valueError = error,
                showLowWarning = showWarning,
                canSave = error == null && intValue != null
            )
        }
    }

    private fun validateValue(value: Int?): String? {
        if (value == null) return null

        return when {
            value < 70 -> "Too low (min 70%)"
            value > 100 -> "Cannot exceed 100%"
            else -> null
        }
    }

    fun saveReading(onSuccess: () -> Unit) {
        val state = _uiState.value
        val value = state.value.toIntOrNull() ?: return

        viewModelScope.launch {
            _uiState.update { it.copy(isSaving = true, saveError = null) }

            try {
                val user = authRepository.getCurrentUser()
                val patientId = user?.linkedPatientId ?: throw IllegalStateException("No patient ID")

                // Create SpO2 observation
                val observation = FhirObservation(
                    id = UUID.randomUUID().toString(),
                    patientId = patientId,
                    type = ObservationType.OXYGEN_SATURATION,
                    value = value.toDouble(),
                    unit = "%",
                    effectiveDateTime = Instant.now().toString(),
                    performerId = user?.userId
                )

                // Save to local store (adds to sync queue)
                localFhirRepository.saveObservation(observation)

                // Play voice acknowledgement
                voicePlayer.playSuccess(ObservationType.OXYGEN_SATURATION)

                _uiState.update { it.copy(isSaving = false) }
                onSuccess()

            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isSaving = false,
                        saveError = e.message ?: "Failed to save"
                    )
                }
                voicePlayer.playFailure()
            }
        }
    }
}

/**
 * UI state for SpO2 screen.
 */
data class SpO2UiState(
    val value: String = "",
    val valueError: String? = null,
    val showLowWarning: Boolean = false,
    val canSave: Boolean = false,
    val isSaving: Boolean = false,
    val saveError: String? = null
)

File: ios/CareLog/CareLog/Vitals/ViewModels/SpO2ViewModel.swift
import Foundation
import SwiftUI
import Combine

/// ViewModel for SpO2 (oxygen saturation) logging screen.
@MainActor
final class SpO2ViewModel: ObservableObject {
    // MARK: - Published Properties

    @Published var value: String = ""
    @Published var valueError: String?
    @Published var showLowWarning: Bool = false
    @Published var isSaving: Bool = false
    @Published var saveError: String?

    // MARK: - Computed Properties

    var canSave: Bool {
        guard let intValue = Int(value) else { return false }
        return validateValue(intValue) == nil
    }

    // MARK: - Dependencies

    private let authService: AuthService
    private let localFHIRRepository: LocalFHIRRepository
    private let voicePlayer: VoiceAcknowledgementPlayer
    private var cancellables = Set<AnyCancellable>()

    // MARK: - Initialization

    init(
        authService: AuthService = .shared,
        localFHIRRepository: LocalFHIRRepository = .shared,
        voicePlayer: VoiceAcknowledgementPlayer = .shared
    ) {
        self.authService = authService
        self.localFHIRRepository = localFHIRRepository
        self.voicePlayer = voicePlayer

        setupValidation()
    }

    // MARK: - Validation

    private func setupValidation() {
        $value
            .dropFirst()
            .sink { [weak self] value in
                guard let self = self,
                      let intValue = Int(value) else {
                    self?.valueError = nil
                    self?.showLowWarning = false
                    return
                }
                self.valueError = self.validateValue(intValue)
                self.showLowWarning = intValue >= 90 && intValue <= 94
            }
            .store(in: &cancellables)
    }

    private func validateValue(_ value: Int) -> String? {
        switch value {
        case ..<70: return "Too low (min 70%)"
        case 101...: return "Cannot exceed 100%"
        default: return nil
        }
    }

    // MARK: - Save

    func saveReading() async {
        guard let spo2Value = Int(value) else { return }

        isSaving = true
        saveError = nil

        do {
            guard let user = await authService.getCurrentUser(),
                  let patientId = user.linkedPatientId else {
                throw VitalError.noPatientId
            }

            // Create SpO2 observation
            let observation = LocalObservation(
                id: UUID().uuidString,
                patientId: patientId,
                type: .oxygenSaturation,
                value: Double(spo2Value),
                unit: "%",
                effectiveDateTime: Date(),
                performerName: user.name
            )

            // Save to local store (adds to sync queue)
            try await localFHIRRepository.saveObservation(observation)

            // Play voice acknowledgement
            voicePlayer.playSuccess(type: .oxygenSaturation)

            isSaving = false

        } catch {
            isSaving = false
            saveError = error.localizedDescription
            voicePlayer.playFailure()
        }
    }
}
